### ISSUE TITLE
model-lab: RTX 3090 fine-tuning lab — faithfulness-gate and correction-intent models

### ISSUE THREAD

**Issue by joshuaswarren** [joshuaswarren/openclaw-engram#1585]
Part of #1572. Standards: #1520. Feeds: #1576 (gate backend), #1581 (intent classifier). Depends on: #1576 shadow mode (training-data harvest), #1584 (downstream eval).

## Problem

Two new classification workloads — the extraction **faithfulness gate** and **correction-intent detection** — will run on every extraction flush. Prompted frontier LLMs work but are slow, non-deterministic, rate-limited, and costly at that call frequency. Both tasks are narrow classification/extraction problems that small fine-tuned models excel at, and the lab hardware (RTX 3090 24 GB, 256 GB RAM, NVMe) is exactly the right size for them.

This issue creates the **model lab**: reproducible recipes to generate training data, fine-tune, evaluate, and serve two small models that Remnic's config can point at — with everything reproducible from manifests and **no datasets or weights ever committed to this repo**.

## Hardware reality (sets the model-size policy)

| Operation | RTX 3090 (24 GB, Ampere — bf16 yes, FP8 no) |
|---|---|
| Encoder classifier fine-tune (DeBERTa-v3-large class, ~0.4B) | Trivial: full fine-tune, minutes–1h |
| Full fine-tune, causal LM | ≤ ~1.5B comfortably |
| LoRA (bf16 base) | ≤ ~8B |
| QLoRA (4-bit base + paged optimizer) | ≤ ~14B (bs 1–2 + grad-accum); ~30B-class technically possible but too slow to iterate — out of policy |
| Inference serving (Ollama / vLLM, 4-bit) | ~14B dense or ~30B-class MoE, one model at a time |

**Policy: target models ≤ 4B for these two tasks** (they are classification, not generation — small models + good data win), with an 8B LoRA escape hatch if evals demand it. Anything larger needs a written justification in the manifest.

### Layout — top-level `model-lab/` (scripts + manifests, NOT an npm workspace package)

```
model-lab/
  README.md                     # hardware envelope, policy table above, quickstart
  common/                       # shared: seeding, JSONL schema, eval runners, HF push helper
  faithfulness-gate/
    generate-data.(ts|py)       # synthetic + harvest pipelines (below)
    train.py                    # single-GPU recipe (TRL/unsloth-class stack)
    eval.py                     # held-out metrics + downstream hooks
    manifest.json               # THE reproducibility artifact (schema below)
  correction-intent/
    (same shape)
.gitignore additions: model-lab/**/data/, model-lab/**/runs/, model-lab/**/*.safetensors, *.gguf
```

Manifest schema (checked into git; the only thing that is): `{task, baseModel: {hfId, revision}, dataRecipe: {generatorGitSha, seed, sources, counts, datasetSha256}, hyperparams, trainedAt, hardware, eval: {heldOut: {...}, downstream: {...}}, artifact: {hfRepo, revision, quantizations}}`. Weights are published to the operator's HF account and referenced by the manifest — git carries recipes and hashes, never blobs (repo ethics rule; checklist §3 — no secrets/PII in any committed file).

### Model 1 — `remnic-faithfulness-gate-v1`

- **Task:** (factText, quote, context) → {entailed, contradicted, unsupported}. Exactly the `FaithfulnessCheckInput` contract from #1576.
- **Base candidates:** (a) DeBERTa-v3-large-MNLI-class encoder with a 3-way head — first choice: cheapest, deterministic, ~1.6 GB serving; (b) a ≤4B instruct causal LM fine-tuned to emit one token class — fallback if the encoder plateaus on long contexts.
- **Training data, three streams:**
  1. **Synthetic perturbation** (the volume driver): take extraction fixtures + generated (quote, fact) pairs; programmatic perturbations create labels — entity swap / negation flip / date shift / quantity change ⇒ `contradicted`; unrelated-quote pairing ⇒ `unsupported`; identity/paraphrase (frontier-model paraphrases of the fact, verified back) ⇒ `entailed`. Perturbations are code, so labels are trustworthy by construction.
  2. **Shadow-log harvest**: #1576 shadow mode records (fact, quote, verdict) with the prompted frontier model as teacher. Harvest is **opt-in, local-only, and documented** — a `model-lab` script reads the operator's own telemetry from their own memoryDir; nothing leaves the machine; the committed dataset hash covers only synthetic + the operator's consented harvest.
  3. **Public NLI warm-start** (optional): standard NLI corpora for pretraining the head before task data.
- **Eval:** held-out F1 per class (target: macro-F1 ≥ 0.9 on synthetic held-out before any downstream claim) + **downstream**: plug into `extraction.faithfulnessModel`, run the #1574 LoCoMo adversarial-category ablation and compare against the prompted-LLM gate on both accuracy and p95 gate latency. The downstream number is the one that matters (rule 55).

### Model 2 — `remnic-correction-intent-v1`

- **Task:** turn text (+ small prior-turn window) → the #1581 `corrections[]` block: {targetHint, correctedAssertion, polarity, confidence} or none. Structured extraction → a ≤4B instruct LM with JSON-schema output is the right shape (an encoder can't emit the spans).
- **Training data:** synthetic conversation generator (personas correcting agents across the full morphology matrix from #1581's fixture taxonomy — reuse those fixtures as the seed grammar) + the anti-example classes (quotes, third-party, hypotheticals) labeled `none`. Frontier teacher labels a sampled slice for validation of the generator's label quality (report agreement in the manifest).
- **Eval:** held-out detection F1 + span quality; **downstream:** MemCorrect (#1584) `false_apply` and `uptake@next` in passive-queue mode vs the prompted classifier.

### Serving + integration

Export merged fp16 + GGUF/AWQ quantizations; serve via Ollama or vLLM on the lab box. Remnic consumes them through existing config only: `extraction.faithfulnessModel` and the #1581 classifier routing pointing at the local openai-compatible endpoint — **zero Remnic-core code changes in this issue**; if a change is needed, it belongs in the consuming child.

## Implementation guide

**PR 1 — lab scaffolding + faithfulness data pipeline.** README (policy table), common utils, gitignore rules, faithfulness `generate-data` with seeded determinism (same seed → same dataset sha256 — test in CI without committing data: generate small-N in a temp dir and assert the hash), perturbation unit tests (each perturbation provably changes/preserves the right label on hand-written cases).

**PR 2 — faithfulness train + eval + manifest.** Train encoder baseline; eval script emits the manifest `eval` block; push to HF; commit manifest. Downstream ablation run + artifact per #1574 protocol.

**PR 3 — correction-intent pipeline + train + eval + manifest.** Same shape; downstream via MemCorrect.

**PR 4 — docs.** `docs/model-lab.md`: how an operator reproduces both models from manifests on their own 3090-class GPU, and how to point Remnic config at them. Every claim traces to a manifest eval block (rule 55).

## Pitfalls

- Python code here follows the global Python standards (3.12+, typed, `requirements.txt`/`setup.sh` local to `model-lab/`) but is NOT part of the npm build/test pipeline — CI must not require a GPU; the only CI hooks are the seeded-generator hash tests (pure CPU, small N).
- Never harvest shadow logs by default; the harvest script requires an explicit `--i-consent-local-data` flag and prints what it will read.
- Teacher-model outputs used for training are "LLM traces" — they live under the gitignored data dir like everything else.
- Version-pin the training stack in the manifest (exact lib versions); a manifest that can't reproduce its own eval numbers is a bug.
- Don't chase 14B models because they fit — iteration speed on a single 3090 is the scarce resource; the policy table is the guardrail.

## Done when

- Both v1 manifests committed with held-out AND downstream eval numbers.
- Remnic on the lab box runs the faithfulness gate + passive correction detection fully locally (no cloud calls) at documented p95 latencies.
- Downstream deltas recorded in bench artifacts (gate quality vs prompted-LLM baseline; false_apply vs prompted baseline).
- Reproduction docs verified by re-running data-gen → hash match.
